presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -207,28 +208,45 @@ public StatementAnalyzer(
         this.session = requireNonNull(session, "session is null");
     }
 
-    public Scope analyze(Node node, Scope scope)
+    public Scope analyze(Node node, Scope outerQueryScope)
     {
-        return new Visitor().process(node, scope);
+        return analyze(node, Optional.of(outerQueryScope));
     }
 
-    private void analyzeWhere(Node node, Scope scope, Expression predicate)
+    public Scope analyze(Node node, Optional<Scope> outerQueryScope)
     {
-        Visitor visitor = new Visitor();
-        visitor.analyzeWhere(node, scope, predicate);
+        return new Visitor(outerQueryScope).process(node, Optional.empty());
+    }
+
+    private void analyzeWhere(Node node, Scope outerQueryScope, Expression predicate)
+    {
+        Visitor visitor = new Visitor(Optional.of(outerQueryScope));
+        visitor.analyzeWhere(node, outerQueryScope, predicate);
     }
 
     private class Visitor
-            extends DefaultTraversalVisitor<Scope, Scope>
+            extends DefaultTraversalVisitor<Scope, Optional<Scope>>
```

Comment:
I wonder if it can be `DTV<Scope, Void>`

Typically when you process a node then you have scopes of its children so you can calculate your scope and return. What are the cases when you need to calculate parent scope which is passed to its children?